File: case_studies/Pgrowth/scripts/datamodel.py
from scipy.integrate import odeint
import numpy as np


def simulate(P, t, γ, λ):
    om = 2.0*np.pi/365.0
    return γ *(1+np.sin(om*t))*P - λ*P**2
	
############################################################

def loglike(theta,obs):
    # step 1: run the model given the parameters to get state
    tobs = obs['t']
    yobs = obs['y']
    
    P = 2.5
    xhat = odeint(simulate, P, tobs, args=theta)[:,0]
    
    # step 2: compute the likelihood (compare state to observations)
    return -np.sum((yobs-xhat)**2)
	
############################################################

T = 365*4
theta = (0.25, 0.1) # γ, λ
P = 2.5
tobs = np.arange(0,T,15)
x = odeint(simulate, P, tobs, args=theta)[:,0]
yobs = x + 0.3 * np.random.normal(size=x.size)

obs = {'t':tobs, 'y':yobs}

############################################################

import matplotlib.pyplot as plt
fig, ax = plt.subplots()
ax.plot(tobs, yobs)
ax.set(title='synthetic observations', xlabel='time (days)', ylabel='P')
None

############################################################

# specify the parameters
theta = (0.2,0.15) # γ, λ

# compute likelihood
L = loglike(theta,obs)
print(L)

############################################################

# gamma

gamma_grid = np.arange(0.1,0.45,0.05)
Lgamma = [loglike((g,0.1), obs) for g in gamma_grid]

fig, ax = plt.subplots()
ax.plot(gamma_grid, Lgamma, marker='o')
ax.set(title='Likelihood Profile: Gamma', xlabel='gamma', ylabel='likelihood')
None

############################################################

# lambda

lambda_grid = np.arange(0.05,0.16,0.01)
Llambda = [loglike((0.25,l), obs) for l in lambda_grid]

fig, ax = plt.subplots()
ax.plot(lambda_grid, Llambda, marker='o')
ax.set(title='Likelihood Profile: Lambda', xlabel='lambda', ylabel='likelihood')
None

############################################################

# 2D likelihood

gamma_grid = np.arange(0.2,0.31,0.01)
lambda_grid = np.arange(0.08,0.125,0.005)

Lacc = np.array([[loglike((g,l), obs) for g in gamma_grid] for l in lambda_grid])

fig, ax = plt.subplots()
cs = ax.contour(gamma_grid, lambda_grid, Lacc, levels=np.linspace(-100,0,11))
ax.clabel(cs, fontsize=10)
ax.set(title='Likelihood surface', xlabel='gamma', ylabel='lambda')
None

###########################################################

from scipy.optimize import minimize

loglike_neg = lambda theta: -loglike(tuple(theta), obs)

res = minimize(loglike_neg, (0.25,0.1), method='BFGS', options={'disp':True})
print('solution: γ={theta[0]:.3f}, λ={theta[1]:.3f}'.format(theta=res['x']))

File: case_studies/Pgrowth/scripts/datamodel.jl
using DifferentialEquations, PyPlot, Distributions, LaTeXStrings, Statistics
using Optim

#########################################################################

simulate(u,p,t) = p[1]*(1+sin(p[2]*t))*u - p[3]*u^2

#########################################################################

function loglike(theta,obs)
    # step 1: run the model given the parameters to get state
    tobs = obs[:,1]
    yobs = obs[:,2]
    u₀ = 2.5;
    tspan = (0.0, maximum(tobs))
    p = [theta[1], 2*π/365, theta[2]]
    prob = ODEProblem(simulate,u₀,tspan,p)
    sol = solve(prob,reltol=1e-6,saveat=15.0)
    xhat = sol.u
    # step 2: compute the likelihood (compare state to observations)
    L = -sum((yobs .- xhat).^2)
    return L
end

#########################################################################

T = 365*4
p = [0.25, 2*π/365, 0.1];
u₀ = 2.5;
tobs = collect(0:15:T)
tspan = (0.0, maximum(tobs))
prob = ODEProblem(simulate,u₀,tspan,p)
sol = solve(prob,reltol=1e-6,saveat=15.0)
yobs = sol.u .+ 0.3 .* rand(Normal(),size(sol.u))

obs = hcat(tobs,yobs);

#########################################################################

fig, ax = PyPlot.subplots()
ax.plot(tobs,yobs)
ax.set_ylabel("P",fontsize=16)
ax.set_xlabel("time",fontsize=16);

#########################################################################

# specify the parameters
theta = [0.2, 0.15] # γ, λ

# compute likelihood
L = loglike(theta,obs)

#########################################################################

# γ
theta = [0.2, 0.15] # γ,om, λ
gamma_grid = collect(0.1:0.05:0.45)
Lγ = zeros(size(gamma_grid))
for i in 1:size(gamma_grid,1)
    theta[1] = gamma_grid[i]
    Lγ[i] = loglike(theta, obs)
end

fig, ax = plt.subplots()
ax.plot(gamma_grid, Lγ, marker="o")
ax.set(title=L"Likelihood Profile: \gamma", xlabel=L"\gamma", ylabel="likelihood");

#########################################################################

# λ
theta = [0.2, 0.15] # γ,om, λ
lambda_grid = collect(0.05:0.01:0.15)
Lλ = zeros(size(lambda_grid))
for i in 1:size(lambda_grid,1)
    theta[2] = lambda_grid[i]
    Lλ[i] = loglike(theta, obs)
end

fig, ax = plt.subplots()
ax.plot(lambda_grid, Lλ, marker="o")
ax.set(title=L"Likelihood Profile: \lambda", xlabel=L"\lambda", ylabel="likelihood");

#########################################################################

# 2D likelihood
theta = [0.2, 0.15] # γ,om, λ
gamma_grid = collect(0.2:0.01:0.3)
lambda_grid = collect(0.08:0.005:0.12)

Lacc = zeros(size(gamma_grid,1),size(lambda_grid,1));
for i in 1:size(gamma_grid,1)
    for j in 1:size(lambda_grid,1)
        theta[1] = gamma_grid[i]
        theta[2] = lambda_grid[j]
        Lacc[i,j] = loglike(theta, obs)
    end
end

########################################################################

fig, ax = plt.subplots()
cs = ax.contour(gamma_grid, lambda_grid, reverse(rotl90(Lacc),dims=1), levels=30)
ax.clabel(cs, fontsize=10)
ax.set(title="Likelihood surface", xlabel=L"\gamma", ylabel=L"\lambda");

function loglike_neg(theta)
    u₀ = 2.5;
    tspan = (0.0, T)
    p = [theta[1], 2*π/365, theta[2]]
    prob = ODEProblem(simulate,u₀,tspan,p)
    xhat = solve(prob,reltol=1e-6,saveat=15.0).u
    L = sum((obs[:,2] .- xhat).^2)
    return L
end

########################################################################

result = optimize(loglike_neg, [0.25,0.1])

